server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java

```diff
@@ -113,6 +113,10 @@ public final class ThreadContext implements Writeable {
      */
     public static final String ACTION_ORIGIN_TRANSIENT_NAME = "action.origin";
 
+    public static final String PLUGIN_EXECUTION_CONTEXT = "_plugin_execution_context";
```

Comment:
I think this is the direction that we should work towards:
 1. The thread context should have caller identity associated with it (Subject or Principal, continuing the work that was done already on `IdentityPlugin`)
 2. The identity comes from the `security` plugin (if installed) and is injected into the thread context using propagators

The examples of the caller identity are:  users, plugins, extensions, system (core). The ContextSwitcher & supported classes should be replaced in favour of `runAs` / `executeAs` impersonation, so each plugin for example could communicate the intent to use different caller identity. 

The goal here, as I see it: have a well defined identity propagation mechanism and an API for plugin / extension authors to use in order to manage identity (the approach with specialized classes like node client etc is not scalable and also easy to circumvent).

Reply:
For what its worth, the 2 clients approach is consistent with how extensions operate. For extensions, the high-level workflow is:

1. On initial handshake, pass a Service Account Token to extension for use when making calls to interact w/ extension's system index
2. When handling an API request on an extension, core will use the IdentityPlugin to issue an on-behalf-of token and forward is to the extension to make calls on-behalf-of the authenticated user

The extension has 2 different clients configured with either 1 or 2.

Extension developer uses the pertinent client when making REST calls to OpenSearch cluster.

Reply:
> the approach with specialized classes like node client etc is not scalable and also easy to circumvent

Curious why you think its not scalable. There would only be 2 different node clients that are meant to be used in different scenarios: 1) node client that executes transport actions in the context of the authenticated user and 2) the pluginNodeClient that is meant to be utilized by plugins that have system indices and is used for transport actions that interact with those system indices.This is a similar pattern done in other places in the OpenSearch ecosystem already where there are 2 clients. Here's one example I encountered last week in ISM where there is a REST client configured with basic auth and one with admin cert: https://github.com/opensearch-project/index-management/pull/1222. The client configured with the admin cert is referred to as the `adminClient`.

I agree on the circumvention via spawning a child thread, but as discussed above changing that would introduce quite a bit of risk since many plugins may rely on spawning threads. Should I file an issue for discussion about how to enforce that plugins spawn threads using a mechanism that maintains the parent context (i.e. `transportService.getThreadPool().generic().submit(thread).get();`).

Reply:
> Curious why you think its not scalable. There would only be 2 different node clients that are meant to be used in different scenarios: 1) node client that executes transport actions in the context of the authenticated user and 2) the pluginNodeClient that is meant to be utilized by plugins that have system indices and is used for transport actions that interact with those system indices. This is a similar pattern done in other places in the OpenSearch ecosystem already where there are 2 clients.

This is exactly the  problem: we used to have 1 client, now (following the same approach) we will have 4 clients. How many clients  (or another components) we would end up maintaining?

> Here's one example I encountered last week in ISM where there is a REST client configured with basic auth and one with admin cert: https://github.com/opensearch-project/index-management/pull/1222. The client configured with the admin cert is referred to as the adminClient.

This is not the same: `AdminClient` is just an interface, it is not another class

Reply:
> Should I file an issue for discussion about how to enforce that plugins spawn threads using a mechanism that maintains the parent context (i.e. transportService.getThreadPool().generic().submit(thread).get();).

I would suggest to focus on caller identity first (https://github.com/opensearch-project/OpenSearch/pull/14630#discussion_r1713676514) and than we could provide mechanism for plugins to use it, for example (very rough idea):

```
transportService.getThreadPool().generic().submit(subject.runAs(thread)).get();).
```